contracts/Curation.sol

```diff
@@ -0,0 +1,529 @@
+        }
+
+        uint256 voteId = voting.newVote(executionScript, "");
+
+        challenges[entryId] = Challenge({
+            challenger: msg.sender,
+            date: getTimestamp(),
+            amount: amount,
+            lockId: lockId,
+            voteId: voteId,
+            dispensationPct: dispensationPct
+        });
+
+        votes[voteId] = Vote({
+            closed: false,
+            result: false,
+            totalWinningStake: 0,
+            votersRewardPool: 0
+        });
+
+        NewChallenge(entryId, msg.sender);
+
+        return entryId;
+    }
+
+    /**
+     * @notice Resolve Challenge for entry with id `entryId`
+     * @param entryId Id of the Application/Challenge
+     */
+    function resolveChallenge(bytes32 entryId) isInitialized public {
+        require(challengeExists(entryId));
+        Challenge storage challenge = challenges[entryId];
+        Application storage application = applications[entryId];
+        Vote storage vote = votes[challenge.voteId];
+
+        require(voting.isClosed(challenge.voteId));
+        vote.closed = true;
+        (vote.result, vote.totalWinningStake,) = voting.getVoteResult(challenge.voteId);
+
+        address winner;
+        address loser;
+        uint256 loserLockId;
+        uint256 amount;
+        if (vote.result == false) { // challenge not accepted (application remains)
+            winner = application.applicant;
+            loser = challenge.challenger;
+            loserLockId = challenge.lockId;
+            amount = challenge.amount;
+
+            // it's still in application phase (not registered yet)
+            if (!application.registered) {
+                application.registered = true;
+                // insert in Registry app
+                registry.add(application.data);
+            }
+        } else { // challenge accepted (application rejected)
+            winner = challenge.challenger;
+            loser = application.applicant;
+            loserLockId = application.lockId;
+            amount = application.amount;
+
+            // Remove from Registry app
+            application.registered = false;
+            registry.remove(entryId);
+        }
+        // compute rewards
+        uint256 reward = amount.mul(dispensationPct) / PCT_BASE;
+        vote.votersRewardPool = amount - reward;
+
+        // unlock tokens from Staking app
+        staking.unlock(application.applicant, application.lockId);
+        staking.unlock(challenge.challenger, challenge.lockId);
+
+        // redistribute tokens
+        staking.moveTokens(loser, winner, reward);
+        staking.moveTokens(loser, address(this), vote.votersRewardPool);
+
+        // Remove used locks
+        delete(usedLocks[application.applicant][application.lockId]);
+        delete(usedLocks[challenge.challenger][challenge.lockId]);
+
+        // Remove challenge, and application if needed
+        if (vote.result == true) {
+            delete(applications[entryId]);
```

Comment:
Removing both application and challenge from storage once resolved is going to make it real hard to get historical data from the app. This would be a major change to how data is stored here, but let's think about it.